feat(authentication): add oauth2 device grant

bastianccm opened the pull request:
Hi,

this change implements the OAuth 2.0 Device Authorization Grant flow for authenticating users without opening a local port.
This has a few advantages, such as the issue described in #717, but also allows remote pipelines (e.g. in gitlab CI) to query for an ephemeral authentication with a remote user.

Before finishing the PR, I'd like to get some feedback about the proposed implementation, specifically:
- `golang.org/x/oauth2` does not really support "non-standard" request, so
  - i am using `net/http` directly to initiate the device authentication
  - i am using `oauth2.SetAuthURLParam` to reuse the `c.oauth2Config.Exchange` method for checking the device-code
~~- I have a couple formatting changes in generated files after running `go generate`, not sure how to handle them?~~

With your feedback provided I'm happy to adjust the implementation, ~~add tests~~, docs and so on.

For the time being this can be tested with:
```
go run . get-token --grant-type device-code --oidc-issuer-url http://localhost:8080/auth/realms/my-issuer --oidc-client-id my-client-id
```
(it has been developed against Keycloak v20 with a client having "OAuth 2.0 Device Authorization Grant" enabled)

I'm looking forward to your feedback!

Helpful links:
Keycloak docs: https://github.com/keycloak/keycloak-community/blob/main/design/oauth2-device-authorization-grant.md
RFC: https://datatracker.ietf.org/doc/html/rfc8628

kstevena commented:
Hi @bastianccm, nice work this was exactly what I was looking for, however I'm in a slightly different context where the provider is Azure AD through an App Registration.
In this context `verification_uri_complete` is empty as there is no option to pass the `user_code` as a parameter of the `verification_uri`. See [Device authorization response
](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code#device-authorization-response)
Then it would be worth to add a check before printing/opening `verification_uri_complete` in case it is empty asking the user to open `verification_uri` and enter the `user_code`.
Doing this small change on my side, I've been able to get this flow working with Azure AD

bastianccm commented:
@kstevena thanks for checking and the feedback!
I changed it:
```go
	if dr.VerificationURIComplete == "" {
		u.Logger.Printf("Please enter the following code when asked in your browser: %s", dr.UserCode)
		u.openURL(ctx, in, dr.VerificationURI)
	} else {
		u.openURL(ctx, in, dr.VerificationURIComplete)
	}
```
I assume this should work as expected then?

kstevena commented:
@bastianccm this works fine with and without `--skip-open-browser` when only the `VerificationURI` is provided.

kstevena commented on pkg/usecases/authentication/device_code/device_code.go:
```diff
@@ -0,0 +1,87 @@
+			tokenSet, err := oidcClient.ExchangeDeviceCode(ctx, dr.DeviceCode)
+			if err != nil {
+				u.Logger.V(1).Infof("err:", err)
+				continue
```
It looks to treat any errors as acceptable until the timeout has expired or the response is successful, while I believe it should only retry if the response code is "400" and the "error" is "authorization_pending".
See: [Device Access Token Response](https://www.rfc-editor.org/rfc/rfc8628#section-3.5)

> The "authorization_pending" and "slow_down" error codes define
>    particularly unique behavior, as they indicate that the OAuth client
>    should continue to poll the token endpoint by repeating the token
>    request (implementing the precise behavior defined above).  If the
>    client receives an error response with any other error code, it MUST
>    stop polling and SHOULD react accordingly, for example, by displaying
>    an error to the user.

bastianccm replied:
I have adjusted the behaviour to only continue on `authorization_pending` and `slow_down` errors, and to slow down by 5 sec if required according to the spec.
Thank you for reviewing and testing!

int128 commented:
Thank you for your awesome contribution.

I wrote a package for the device flow, because it would be nice to keep the implementation of kubelogin simple.
https://github.com/int128/oauth2dev

Your implementation looks very nice and I think okay to merge now. 
I would like to consider the migration to the package in the future.

bastianccm commented:
@int128 thank you for the feedback, if you like I can migrate to your package right away, then we have even less code to review :) (I wasn't aware of your implementation)

int128 commented:
@bastianccm Could you try using my package in this pull request? I'd like to help you if there is some issue.

bastianccm commented:
@int128 I created a new PR with your package, it is blocked by https://github.com/int128/oauth2dev/issues/11 right now - I'll try to fix it and submit a PR

bastianccm commented:
@int128 PR updated

The oauth2dev package does not support client authentication, which matches the spec in RFC 8628 (there is no client secret mentioned).
Keycloak (which I use for testing) would support that, so I am not exactly sure if we should too? For now it works if no client secret is specified, so generally we are fine.

kstevena commented:
@bastianccm with the recent changes you made it stopped working on my side with following error:
```
I1219 11:00:38.934448   25484 device_code.go:40] finished the oauth2 device code flow
error: get-token: authentication error: device-code error: unable to exchange device code: id_token is missing in the token response: &{THE_ACCESS_TOKEN Bearer  2022-12-19 12:00:37.934293219 +0100 CET m=+3615.635903323 <nil>}
```
While looking at the response received it contains:
`{"token_type":"Bearer","scope":"email openid profile User.Read","expires_in":3599,"ext_expires_in":3599,"access_token":"THE_ACCESS_TOKEN","id_token":"THE_ID_TOKEN"}`

It looks `ExchangeDeviceCode` tries to verify the token without passing the `id_token` from the response:
```
	return c.verifyToken(ctx, &oauth2.Token{
		AccessToken:  tokenResponse.AccessToken,
		TokenType:    tokenResponse.TokenType,
		RefreshToken: tokenResponse.RefreshToken,
		Expiry:       tokenResponse.Expiry(),
	}, "")
```
while `verifyToken` expect to find `id_token` in it:
```
func (c *client) verifyToken(ctx context.Context, token *oauth2.Token, nonce string) (*oidc.TokenSet, error) {
	idToken, ok := token.Extra("id_token").(string)
	if !ok {
		return nil, fmt.Errorf("id_token is missing in the token response: %s", token)
	}
```

bastianccm commented:
@kstevena you are correct, it slipped thru while testing.

https://github.com/int128/oauth2dev/pull/13 addresses the handling of id_token (and other claims), which we need to make this change working.
I updated this branch and replaced the module with that PR for testing.

int128 commented on go.mod:
```diff
@@ -10,14 +10,15 @@ require (
 	github.com/google/go-cmp v0.5.9
 	github.com/google/wire v0.5.0
 	github.com/int128/oauth2cli v1.14.0
+	github.com/int128/oauth2dev v0.0.0-20221220102744-82ff1a972401
```
I will create a stable tag after this PR.

int128 commented on pkg/oidc/client/client.go:
```diff
@@ -12,14 +12,27 @@ import (
 	"golang.org/x/oauth2"
 )
 
+// DeviceRequest encodes an oauth2 device authorization request
+type DeviceRequest struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationURI         string `json:"verification_uri"`
+	VerificationURIComplete string `json:"verification_uri_complete"`
+	ExpiresIn               int    `json:"expires_in"`
+	Interval                int    `json:"interval"`
+}
+
```
Can we remove this? It seems not used.

int128 commented on pkg/usecases/authentication/device_code/device_code.go:
```diff
@@ -0,0 +1,67 @@
+package device_code
```
Nits, it is recommended to use a package name without `under_scores`.
https://go.dev/blog/package-names

```suggestion
package devicecode
```

int128 reviewed (commented):
Thank you for nice contribution.
